### ISSUE TITLE
Insane CPU usage under linux with python ext, with pylance even worse

### ISSUE THREAD

**Issue by crankedguy** [microsoft/pylance-release#2382]
Hello,

I recently realized really insane CPU usage under Linux (Fedora) with python extension enabled.
When I enable pylance it is even worse.
If I disable both extensions I can have several Code windows open and my CPU load is 0.8 - 1.5% (24 core Threadripper)
If I enable python extension I have one process with more or less permanent 100% core load, if I enable Pylance on top of it there are 2 processes with high load shown. If I have a few Code windows open then my total CPU load is at about 8-10% which is insane, and costs me a lot of money over time on top of it.

Information I can provide 
Without pylance
![Screenshot_20220214_111906](https://user-images.githubusercontent.com/34091479/153847060-0232715e-0b15-4d27-a77b-53d7600743a3.png)
With pylance
![Screenshot_20220214_112021](https://user-images.githubusercontent.com/34091479/153847160-8424a0a1-03ed-4c7b-967c-2ff714c0edea.png)
Code version
![Screenshot_20220214_112647](https://user-images.githubusercontent.com/34091479/153847216-227d3a17-7963-4033-a046-a3afb49e45d5.png)
python settings in config:
```   
    "python.linting.flake8Enabled": true,
    "python.linting.flake8Args": [
        "--select=E,F,W,C90",
        "--ignore=E203,E266,E501,W503,E712,E711,E713,W504,W605",
        "--max-line-length=120",
    ],
    "python.formatting.provider": "yapf",
    "python.languageServer": "Default",
    "python.formatting.yapfArgs": [
        "--style",
        "{based_on_style: google, column_limit:120 }"
    ],
```

You have any idea how to handle that?

**Comment by billytheme** [comment]
I have also noticed this but. My Python extension is causing me any grief, unlike yours, but my Pylance uses 100% of one core for around 5 minutes each time I save a file. This only started about a week ago, unsure if there was a new version pushed to VSCode?

**Comment by erictraut** [comment]
@billytheme, we fixed a bug a few days ago that is the likely cause of the problem you're seeing. We typically release a new version of Pylance each Wednesday, so let's see if this week's version addresses your issue. If not, we'll investigate further.

**Comment by crankedguy** [comment]
@omicscoutspweber : Thanks for the hint with 2022.1.3 
The situation seems a lot better there with only the occasional 100% spike of pylance, but way better and cheaper to what is happening with the actual version.

**Comment by crankedguy** [comment]
but still, also with that version... 2 cores/2processes 100% - for a prolonged time (up to minutes) - when editing a bigger file (> 4k lines), this seems odd to me

**Comment by billytheme** [comment]
Hi @erictraut 

Which version number is that specifically? I only the latest version as 2022.2.1, released last week. Is this week's updated version out yet or not?

**Comment by debonte** [comment]
> Is this week's updated version out yet or not?

Not yet. We typically ship Pylance on Wednesdays.

**Comment by mhillebrand** [comment]
v2022.2.3 fixed this issue for me

**Comment by debonte** [comment]
@crankedguy can you check whether 2022.2.3 fixes this on your machine as well?

**Comment by crankedguy** [comment]
@debonte Sadly did not fix a thing. What has changed is that there is not a permanent 100%CPU load anymore, but this is only sounding nice. It is cosmetical... Because as soon as I touch a file, meaning putting one char into it and save it, I have two cores 100% load for over 2 minutes. So the issue of being expensive stays because this is what is happening the whole day, in a permanent manner.
There is no remote environment involved like with @mhillebrand. Working on local files, not even over NFS but real local SSDs.
The difference to 2022.1.3 is a plus of over 1:30 more in 200% CPU load or in pct time we are talking 325% CPU load vs 2022.1.3
This should really be investigated from my point of view

**Comment by mgsnuno** [comment]
@debonte I faced this issue yesterday after updating vscode, but this morning with the latest update of PyLance all back to normal. just reporting my status.

**Comment by erictraut** [comment]
@crankedguy, you mentioned that you see an increase in CPU load when you insert a new character and save the file. Can you confirm that the CPU increase starts only after you save the file and not immediately after inserting a new character? That's an important clue. If the problem occurs only after saving a file, that leads me to believe it's an issue with the indexer. You could try disabling it by setting `python.analysis.indexing` to false. Does that mitigate the problem?

**Comment by erictraut** [comment]
@crankedguy, it sounds like the issue you're seeing is unrelated to the one that was fixed in v2022.2.3. We'll need more details from you to diagnose your issue. By any chance, is your code available in a public github repo? The most effective way for us to diagnose an issue like this is to have a local repro case. If that's not possible, then we might be able to figure it out from logs and directed experiments to help narrow down the issue.

**Comment by crankedguy** [comment]
@erictraut Thanks for your reply. No sadly it is not available in a public repo. It is work in progress and in parts still raw, in parts code from another project of me...
But we have another "problem" now. I refactored quite some code yesterday until late. Today I checked now and we are down from over 2 minutes to like 20-30 seconds. I have no commit between the two, the last commit is way too far away. As said I am still in a raw state that is why there are not many commits etc. But there had to be something in that code that the extension could not handle properly or made it bug out.

But there is still something in the code, as it is still way slower/higher load than in other files. 
I don't want to mess around with you guys. For me 20-30 seconds 2x100% CPU load for typing a character is still a lot, probably too much, but if you say this is not worth investigating anymore I have to live with it. Maybe it goes back to extremely bad, I am not finished as said.

If you say we look further into it, which I would appreciate, I will provide you with all the information I can give/testing I can do.

**Comment by erictraut** [comment]
I'm interested in investigating further, but if you can't share the code with me, you'll need to do most of the work to narrow down the problem. The first thing I'd do in a situation like this is to identify the function and expression that is contributing to the slowdown. This can be done by commenting out sections of the file until the performance issue goes away. 

The other thing you can do is enable the following settings:
```
  "python.analysis.logLevel": "Trace",
  "python.analysis.logTypeEvaluationTime": true,
```

This will cause pylance to emit more verbose log data to the output tab. This may give you some clues to help narrow down the issue.

Once you've isolated the minimal code sample that causes the problem, I'd need you to describe it or (preferably) post it to a gist or sample repo.

**Comment by crankedguy** [comment]
@erictraut I can share the code with you but I am not keen to open a public repo right now for it. Should I send it to you as a tar/zip?

**Comment by erictraut** [comment]
Could you give me temporary read access to your github repo? Or extract out the relevant code to a separate github repo that you give me access to temporarily?

**Comment by crankedguy** [comment]
1. It is not even on gtihub, not even as private repo, it is only local. 
2. I only upload repos that I consider in a more or less release state for myself.
3. I normally use bitbucket :) :)
4. But I will go on and upload it and give you acess to it as I am interested in that myself. I'll get back to you, I start right now. thanks!

**Comment by crankedguy** [comment]
I copied it and created a new repo for it for now, you should have gotten an access invitation

**Comment by erictraut** [comment]
@crankedguy, thanks, I was able to repro the problem you're seeing. I think it's unrelated to the other issues in this thread, so I've opened a new issue [here](https://github.com/microsoft/pyright/issues/3081).

**Comment by bschnurr** [comment]
This issue has been fixed in version 2022.3.0, which we've just released. You can find the changelog here: [CHANGELOG.md](https://github.com/microsoft/pylance-release/blob/main/CHANGELOG.md#202230-3-March-2022)
